ls1intum/Artemis
# Show online status of team members in text and modeling editor

### Checklist

### Motivation and Context
Team-based text and modeling exercises require the students to be aware of what their team members are doing since they are all working on the same artifact. This PR is a first step towards that: On the text editor page and on the modeling editor page, students can now see if they are connected to the server and who else of their team is currently looking at the artifact.

### Description
1. Added `ParticipationTeamWebsocketService.java` (handles subscribe and unsubscribe events for team participations; lets a user know on subscribe who else is currently subscribed to the topic, i.e. watching the same participation)
2. New component: Team Students Online List (vertical chat-like list of users with online indicator dot in front of the names; encapsulated websocket functionality and uses the service above to receive information about who if the team is currently looking at the same page)
3. New component: Team Participate Info Box (combines personal connection status, max score and team students online list into an information box; it is usually placed on the right side of the problem statement; it supports bounded or unbounded sticky positioning when scrolling so that it stays visible when possible)
4. Adapt text editor page and modling editor page to display the `Team Participate Info Box` when the exercise is team-based. If it is a normal, individual-based exercise, the appearance stays the same as it was until now (except for a small column optimization where the right max score only takes as much space as it actually needs).

### Steps for Testing
1. Log in to Artemis
2. Create a team-based text and a modeling exercise ([how?](https://github.com/ls1intum/Artemis/pull/1282#issuecomment-605954482))
3. Create at least one team with at least two students for each exercise
4. Start both exercises as anyone of the students from the teams
5. Log in with two students of the same team at the same time (use incognito mode or a different browser)
6. For both, go to the text or modeling editor page
7. Check that the team list is shown
8. Check that the online status behaves correctly

### Screenshots
**Note:** Updated screenshots in the comments.

1. Team Text Exercise: Right side has connection status and list of team members with their online status (team member list always shows yourself first and the rest alphabetically). A green dot means that the other person has the same page open.
![OnPaste 20200406-185351](https://user-images.githubusercontent.com/7109225/78584375-6f5ba480-7838-11ea-839c-9f4e984c5628.png)

2. When scrolling, the information is presented in a sticky box with a light box-shadow. For text-exercises, it scrolls along until the very bottom of the page (so it might slightly overlap the textarea field for smaller browser widths).
![OnPaste 20200406-185458](https://user-images.githubusercontent.com/7109225/78584380-6ff43b00-7838-11ea-9a72-39133ee1fa6d.png)

3. On mobile, this info is placed between the problem statement and the textarea.
![OnPaste 20200406-185547](https://user-images.githubusercontent.com/7109225/78584383-708cd180-7838-11ea-86af-65cf15691fa2.png)

4. Team Modeling Exercise: Same familiar display as for text exercises.
![OnPaste 20200406-185612](https://user-images.githubusercontent.com/7109225/78584386-71256800-7838-11ea-9b97-116331c62af5.png)

5. The sticky box on scrolling is bounded here, meaning it stops at a certain point (when the bounding container, i.e. the problem statement, ends). This is done so that the modeling editor is not overlapped since that could be really annoying for the user. Also since the height of the modeling editor is more or less fixed compared to the flexible textarea for text exercises, the team online info can be kept in the viewport if intended by the user.
![OnPaste 20200406-185640](https://user-images.githubusercontent.com/7109225/78584390-71bdfe80-7838-11ea-9868-d6a32eee6dc8.png)

6. Same behavior as for text exercises on mobile.
![OnPaste 20200406-185702](https://user-images.githubusercontent.com/7109225/78584391-71bdfe80-7838-11ea-9de9-6bba39c32a6c.png)
